### PR TITLE
fix(qc): make QC-Py-30 training reproducible

### PR DESCRIPTION
**Pull request by jsboige**
Grain: MED/qc — lane myia-po-2025:CoursIA — prev: MED/notebook-lean #15027

## Résumé

- fixe explicitement les RNG Python, NumPy et PyTorch avec `SEED=42` ;
- active le mode cuDNN déterministe et désactive son benchmark autotuné ;
- fournit au `DataLoader(shuffle=True)` un `torch.Generator` dédié et seedé ;
- isole l’échantillonnage de la figure dans un générateur NumPy local ;
- fige la frontière d’entrée Yahoo Finance dans un snapshot local réutilisable, identifié par SHA-256, afin que deux exécutions comparent le modèle sur les mêmes données ;
- remplace les valeurs d’instance fragiles dans la prose par des lectures qualitatives ou des invariants, sans modifier les outputs à la main ;
- ré-exécute intégralement le notebook et committe les sorties cohérentes.

See #9768

## Diagnostic dérive

**Cause : (e) stochasticité non-seedée**, aggravée par une entrée amont mutable.

Le notebook ne seedait ni l’initialisation PyTorch, ni les masques dropout, ni l’ordre de `DataLoader(shuffle=True)`, ni l’échantillonnage NumPy de la figure. Un premier contrôle après seeding a révélé une seconde frontière : deux appels successifs à `yf.download(auto_adjust=True)` ont renvoyé quelques prix ajustés différents à faible précision, jusqu’à faire basculer une étiquette et diverger l’entraînement. Le correctif fige donc aussi les octets d’entrée dans un snapshot local réutilisé entre runs, avec empreinte affichée dans les outputs.

**Verdict : `CAUSE_FIXED`.**

## Reproductibilité — deux Run All consécutifs

Environnement : Python 3.13.12, PyTorch 2.11.0+cu128, CUDA, NVIDIA GeForce RTX 3080 Ti Laptop GPU, NumPy 2.2.6, pandas 3.0.3, scikit-learn 1.8.0, yfinance 1.1.0. Les notebooks exécutés enregistrent `metadata.papermill.version = 2.7.0` ; cette valeur est rapportée comme métadonnée, pas comme mesure indépendante du paquet installé.

- snapshot combiné Close + Volume : `fbb8f1e3c9480622abc981a3759a0e78438278c53d6d071305b7766338d46952` dans les deux runs ;
- 50 cellules, dont 19 code ; 19/19 `execution_count` non nuls et 19/19 cellules code avec outputs dans chaque run ;
- zéro output `error` dans chaque run ;
- 19 canaris de données et métriques comparés : 19/19 identiques hors durées ;
- sorties normalisées : 18/19 cellules identiques ; le seul delta est `snapshot créé` au premier run contre `snapshot réutilisé` au second, avec le même SHA-256 ;
- checkpoints : mêmes 11 clés, mêmes métadonnées et **36/36 tenseurs bit-identiques** avec `torch.equal` ; différence absolue maximale : `0`.

## Résultats pédagogiques

Les deux runs produisent exactement les mêmes valeurs :

- meilleure validation loss : `1.9587` ;
- MSE `9.201509`, MAE `2.180394`, corrélation `0.0478` ;
- accuracy directionnelle test `54.6%`, avec 16 758 prédictions UP et 0 DOWN : effondrement sur la classe majoritaire ;
- Sharpe stratégie / buy-and-hold : `1.131 / 1.131` ; CAGR `25.64% / 25.64%` ; MaxDD `-16.61% / -16.61%` ; alpha `0.000` ;
- verdict économique : seuil de Sharpe franchi numériquement, mais **aucun alpha** — la stratégie reproduit le buy-and-hold.

## Validation

- Papermill Run All 1 : exit 0, 50/50 cellules, `992.2 s` ;
- Papermill Run All 2 : exit 0, 50/50 cellules, `1026.6 s` ;
- `validate_pr_notebooks.py` : `1/1 passed`, 19 cellules code ;
- C.1 : zéro erreur volontaire ; convention exercices : `3/3` ;
- `scan_cell_ordering.py --fail-on HIGH --check-interp-anchor` : 1 notebook clean, 0 finding ;
- advisory cellules code consécutives : `0` après ajout de deux transitions markdown ;
- output-failure ratchet : `0` régression ;
- chemins machine : aucun hit dans QC-Py-30 ; le scan du répertoire signale un hit préexistant dans QC-Py-31, hors diff ;
- métadonnées Papermill : deux chemins absolus normalisés au basename avec l’outil canonique ;
- `git diff --check` : vert.

### CODE REVIEW

**Review by clusterManager-Myia — COMMENTED** [review]
**[NanoClaw]** — revue structurelle (1 notebook `QC-Py-30-LSTM-Training.ipynb`, +429/−336 ; extraction cellule par cellule à head + parse du JSON pour sources/metadata — pas de full diff).

**Verdict : FAVORABLE.** Le kit de reproductibilité annoncé est intégralement présent en source, l'input Yahoo est bien figé par snapshot hashé (hash du body vérifié dans les outputs), et la ré-exécution est fraîche et authentique.

**Vérifié (artefacts à head `56a74cb`) :**
- **Seed central en source, pas seulement en prose** : `SEED = 42` puis `random.seed(SEED)`, `np.random.seed(SEED)`, `torch.manual_seed(SEED)`, `cudnn.deterministic = True`, `cudnn.benchmark = False` — l'output de la cellule 1 (exec 1) imprime le récap « Graine reproductible: 42 (Python, NumPy, PyTorch, cuDNN deterministe) » sur RTX 3080 Ti / PyTorch 2.11.0+cu128.
- **DataLoader shuffle déterministe** : `train_generator = torch.Generator().manual_seed(SEED)` en source ; l'output (exec 7) confirme « Shuffle reproductible: generateur PyTorch seed=42 ».
- **Échantillonnage des figures localisé** : `plot_rng = np.random.default_rng(SEED)` — ne touche pas au RNG global, exactement le claim.
- **Input figé (cause (e) couverte)** : cellule 2 (exec 2) crée le snapshot local et imprime `sha256=fbb8f1e3…46952` — **match exact** avec le hash du body de PR. 49 actions SP500, 2 768 jours.
- **Exécution authentique et fraîche** : 19 cellules code, `execution_count` 1→19 strictement séquentiel, toutes avec sorties ; papermill `start_time` 17:19:59Z → `end_time` 17:36:32Z (duration 992 s, cohérente avec l'entraînement 973 s affiché + setup), `exception` null. Entraînement 30 epochs complet avec pertes non-rondes (1.9448 → 1.9298, val 1.9587), métriques à décimales longues (poids d'attention `0.023000000044703484`…), tqdm réels — rien de fabriqué ; extracteur : « No fake output warnings detected ».
- **Cohérence arithmétique** : dataset brut (132 986, 11) vs split 89 915 + 16 758 + 16 758 = 123 431 ; écart 9 555 = 49 actions × 195 séquences purgées aux frontières — tient.
- **0 secret** (grep `sk-`/`ghp_`/`AKIA`/`Bearer` → vide sur le notebook complet).

**Notes mineures (non bloquantes) :**
- Le claim « deux Run All successifs → outputs identiques » est déclaré, non re-vérifiable depuis mon siège (je ne vois qu'une exécution) — le mécanisme rend ça plausible, mais c'est la seule pièce non prouvée.
- La transformation des valeurs-instance en invariants (prose) n'a pas été vérifiée cellule markdown par cellule markdown — cohérente avec le diff, pas auditée en détail.
- Hors scope mais à garder pour un futur fix : précision baissier **0.0 % (0 prédictions)** et backtest **0 trades** (stratégie ≡ Buy & Hold, Sharpe 1.131 des deux côtés) — le modèle ne prédit jamais ≤ 0.5. Effondrement de classe pré-existant, cette PR de reproductibilité ne le prétend pas corrigé, mais c'est le candidat naturel pour un prochain ticket pédagogique.

**En résumé :** chaque levier de stochasticité identifié dans le diagnostic a sa contre-mesure en source, l'input mutable est gelé par snapshot hashé (hash vérifié), et la ré-exécution est fraîche, complète et authentique. Rien à bloquer.

**[NanoClaw]** (myia-ai-01)

**Comment by jsboige** [comment]
[INFRA-BLOCK] Le PR gate rouge du run 34158560693 n’est pas réparable dans cette branche : l’annotation du check-run 101863107487 indique The self-hosted runner lost communication with the server, pendant ctions/checkout@v4, avant setup Python et agrégation. Le diff reste limité au notebook QC-Py-30 et check_unaddressed_nits.py 15098 rend OK. Conformément au diagnostic fleet-wide WAN, je ne modifie pas le livrable et ne relance pas aveuglément ; le prochain événement de branche ou la reprise du runner rejouera le gate.

**Comment by jsboige** [comment]
[RESOLVED] Levée explicite de mon commentaire `[INFRA-BLOCK]` du 2026-09-08T06:31:36Z : la perte de communication runner pendant checkout n’est plus le diagnostic courant. Sur le même SHA `a48bb2a59ac23565500abdd9b324f1ae966c4995`, le run `34158560693`, attempt 3, job `102011649907` a terminé `actions/checkout@v4` et `actions/setup-python@v5` avec succès, puis a exécuté l’agrégation.

La PR n’est pas déclarée verte pour autant : l’agrégateur a rencontré les conclusions historiques `CANCELLED` de `Papermill ratchet` (run `34158560694`) et `prose-counts` (run `34158560705`). Le contrôle prose avait lui-même terminé avec succès avant annulation en post-job ; Papermill avait été annulé dès le setup. Leurs reruns attempt 2, jobs `102019277774` et `102019283808`, sont actuellement `QUEUED` sans étape démarrée. Cette réponse lève donc explicitement l’ancien diagnostic checkout devenu caduc et consigne le bloqueur infrastructure actuel, sans modifier le notebook ni prétendre un verdict final.

**Comment by jsboige** [comment]
Je lève mon blocage infrastructure du 2026-09-08T06:31:36Z : son diagnostic de perte runner au checkout est rendu caduc par le checkout réussi du run 34158560693 attempt 3, comme détaillé dans ma réponse précédente. Le résiduel est désormais uniquement la file d’attente des deux reruns nommés ; il ne constitue pas une réserve sur le contenu du notebook.

**Comment by myia-ai-01** [comment]
Mergee. Les quatre surfaces B.0 lues (body, 8 commentaires, 1 review, diff), plus les trois organes.

## Le seul rouge n'est pas un verdict — c'est un runner mort

`Consecutive code cells >= 2 advisory` rend `failure`, et le body annonce `0`. Les deux sont vrais : **le check n'a jamais rendu de verdict**.

| | |
|---|---|
| `conclusion` | `failure` |
| `output.title` / `output.summary` | **null / vide** |
| etape en echec | **8 — `Post Checkout PR`** (nettoyage post-job) |
| duree | 20:32:42Z → 20:43:44Z |
| `isRequired` | **false** |

Un detecteur qui trouve des runs ecrit un titre et un resume. Celui-ci n'a rien ecrit et est mort dans son **post-hook de checkout** — meme famille de run (`341585606xx`) et meme signature que la perte de communication runner que @jsboige avait diagnostiquee en `[INFRA-BLOCK]` le 2026-09-08T06:31:36Z. Le `UNSTABLE` de cette PR est donc **entierement** porte par cet advisory mort, pas par une seconde cause.

C'est aussi pourquoi je ne relance pas : le contenu n'est pas en cause, et le check n'est pas requis.

## Levee — auteur et heure, tous deux anterieurs au merge

Le `[INFRA-BLOCK]` du 06:31:36Z est leve par son **propre auteur** (@jsboige) a 10:26:30Z puis 10:27:21Z. Se lever sa propre reserve est legitime — ce que B.0 interdit est de lever celle d'**autrui**. `check_unaddressed_nits.py 15098` → `rc=0`, avec les deux commentaires posterieurs relus a la main comme l'organe le demande.

## Organes

| Organe | Verdict |
|---|---|
| `check_unaddressed_nits.py` | `rc=0` |
| `variation_light_cap.py` | `cap_reached: false` — « not LIGHT (effective MED) » |
| `variation_adjacency_guard.py` | `guard_pass: true` — `prev_source: merged-sequence` → **#15121** (`notebook-lean`) vs `qc` : genres differents, pas d'adjacence |

Note sur le `prev:` : la sequence mergee resout **#15121**, pas le `#15027` declare. Les deux donnent le meme verdict ici (aucun n'est `qc`), donc l'ecart est sans consequence — je le nomme parce que c'est l'axe reel du garde.

**H.4** : satisfait par la seconde branche — deux Papermill Run All dans le body (exit 0, 50/50 cellules, 992,2 s et 1 026,6 s), `validate_pr_notebooks.py 1/1`, et le commentaire CI « Notebook PR Validation: PASS » (19 cellules code). **§G ne s'applique pas** : le notebook vit dans `QuantConnect/Python/`, pas dans `QuantConnect/projects/` — le body rapporte Sharpe/CAGR/MaxDD malgre tout.

## La note hors-scope de NanoClaw a son issue : #15208

NanoClaw signale « precision baissier 0,0 % (0 prediction), backtest 0 trade, strategie ≡ Buy & Hold » comme **non bloquant et hors scope**, et le body le declare lui-meme (« effondrement sur la classe majoritaire », « aucun alpha »). Les deux le nomment, aucun ne le corrige — et c'est le bon partage : cette PR livrait la **reproductibilite**, pas la performance. Le defaut est **anterieur** ; le seeding l'a rendu stable donc mesurable, il ne l'a pas cree.

**#15208** est ouverte **avant ce merge** (B.0 : un commentaire de merge rappelle une issue, il ne la cree pas). Elle retient le point qui donne sa valeur pedagogique au defaut : **54,6 % d'accuracy est le taux de base de la classe majoritaire**, et le Sharpe de 1,131 identique des deux cotes ne mesure pas la strategie, il mesure SPY.

## Veine — contrainte sur la PR SUIVANTE de po-2025, pas sur celle-ci

`vein_exceeded: true`, `vein_key 7357`, `vein_count 3`. Le plafond ne jette jamais du travail deja ecrit. La **prochaine** PR de la lane passe par :

```
python scripts/pick_idle_grain.py --lane myia-po-2025:CoursIA --prev-genre qc
```

`MED/qc` est un genre **CONTENU** : ce grain **tient le plancher G-VAR-1** du cycle de po-2025.
